### ISSUE TITLE
Major Weight update bug detected

### ISSUE THREAD

**Issue by athus1990** [keras-team/keras#12805]
Simple model that trains a simple 1 batch object:

``` import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten, Activation
from keras.layers import Conv2D, MaxPooling2D
from keras.optimizers import SGD
import copy


def load_data():
    (x_train, y_train), (_, _) = keras.datasets.cifar10.load_data()
    x_train = x_train[:, :5, :5, :]  # (50000, 5, 5, 3)
    y_train = np.squeeze(y_train)
    y_train_1_hot = np.empty(shape=(y_train.shape[0], 10), dtype=np.int32)
    for value in range(0, y_train.shape[0]):
        yarg = y_train[value]
        yhot = [0] * 10
        yhot[yarg] = 1
        y_train_1_hot[value] = yhot
    return (x_train, y_train_1_hot)

xtrain, ytrain = load_data()
print(xtrain.shape)
print(ytrain.shape)
one_sample_input = xtrain[0:1]
#one_sample_input = np.expand_dims(one_sample_input, 0)
one_sample_output = ytrain[0:1]
#one_sample_output = np.expand_dims(one_sample_output, 0)
model = keras.models.Sequential()
model.add(Conv2D(2, (3, 3), activation=None, use_bias=False, trainable=True, input_shape=(5,5,3)))
#model.add(Activation('relu'))
model.add(Flatten())
model.add(Dense(10, use_bias=False,activation='softmax'))
rndw = [np.random.rand(*x.shape)for x in model.get_weights()]
model.set_weights(rndw)
##model weights before training
x0=model.layers[0].get_weights()[0] 
x1=model.layers[2].get_weights()[0]
#model.compile(loss='categorical_crossentropy', optimizer=keras.optimizers.SGD(lr=1,momentum=0,nesterov=False), metrics=['accuracy'])
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(x = one_sample_input, y = one_sample_output, batch_size=2, epochs=2, verbose=1)
##model weights after training
a0=model.layers[0].get_weights()[0]
a1=model.layers[2].get_weights()[0]
print(np.max(a0-x0),np.max(a1-x1)) 

``` 

**prints 0.0 and 0.0
Therefore the model weights do not change!!**

However , Removing smftmax from the last layer:
 
**prints Non zero
Therefore the model weights change!! Softmax is casing some weird issue**

``` 
import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten, Activation
from keras.layers import Conv2D, MaxPooling2D
from keras.optimizers import SGD
import copy
def load_data():
    (x_train, y_train), (_, _) = keras.datasets.cifar10.load_data()
    x_train = x_train[:, :5, :5, :]  # (50000, 5, 5, 3)
    y_train = np.squeeze(y_train)
    y_train_1_hot = np.empty(shape=(y_train.shape[0], 10), dtype=np.int32)
    for value in range(0, y_train.shape[0]):
        yarg = y_train[value]
        yhot = [0] * 10
        yhot[yarg] = 1
        y_train_1_hot[value] = yhot
    return (x_train, y_train_1_hot)

xtrain, ytrain = load_data()
print(xtrain.shape)
print(ytrain.shape)
one_sample_input = xtrain[0:1]
#one_sample_input = np.expand_dims(one_sample_input, 0)
one_sample_output = ytrain[0:1]
#one_sample_output = np.expand_dims(one_sample_output, 0)
model = keras.models.Sequential()
model.add(Conv2D(2, (3, 3), activation=None, use_bias=False, trainable=True, input_shape=(5,5,3)))
#model.add(Activation('relu'))
model.add(Flatten())
model.add(Dense(10, use_bias=False))
rndw = [np.random.rand(*x.shape)for x in model.get_weights()]
model.set_weights(rndw)
##model weights before training
x0=model.layers[0].get_weights()[0] 
x1=model.layers[2].get_weights()[0]
#model.compile(loss='categorical_crossentropy', optimizer=keras.optimizers.SGD(lr=1,momentum=0,nesterov=False), metrics=['accuracy'])
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(x = one_sample_input, y = one_sample_output, batch_size=2, epochs=2, verbose=1)
##model weights after training
a0=model.layers[0].get_weights()[0]
a1=model.layers[2].get_weights()[0]
print(np.max(a0-x0),np.max(a1-x1)) 
``` 

**System information**  
- Have I written custom code (as opposed to using example directory):  NO
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Linux Ubuntu 14.04
- TensorFlow backend (yes / no):  yes
- TensorFlow version:  1.4
- Keras version:  2.1.3
- Python version:  2.7
- CUDA/cuDNN version:  8.1
- GPU model and memory:  Titan X

**Comment by kashyapraval** [comment]
seems like an error in evaluation of softmax. Please check if the dtype is float32. I doubt even float64 may help. You could have a look at what the values are towards the end layers.
Also, I don't believe you are attempting to train on parts of images. Please check your splicing. for example, check the shape of 1st train sample. Instead of 3 frames, you get data from 5 images.

**Comment by kashyapraval** [comment]
for example, the softmax for 1st training is evaluated as [inf/inf, ..., inf/inf] which results in the nan array.

**Comment by athus1990** [comment]
@kashyapraval the funny thing is it is float32. More importantly this is a test example to check with just one batch size if weights gets updated after 1 epoch.


Input image is of dimension 1,5,5,3 and output label is of dimension 1,10. the dimensions I believe are correct ( batch size, h,w ,channel)

Even if I don't randomly init weights and do the default weights that keras uses. The same issue happens.

**Comment by hondaathma** [comment]
Moreover there are no inf or Nan values anywhere...even before or after training 1 epoch

**Comment by athus1990** [comment]
Moreover there are no inf or Nan values
